### PR TITLE
Implement basic text rendering support in CanvasKit backend

### PR DESCRIPTION
**Pull request by hterkelsen**
This adds basic text rendering support using the "SkParagraph" module in Skia.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+
+    if (_fontFamilyOverrides.containsKey(fontFamily)) {
+      fontFamily = _fontFamilyOverrides[fontFamily];
+    }
+    List<String> fontFamilies = <String>[fontFamily];
+    if (fontFamilyFallback != null) {
+      fontFamilies.addAll(fontFamilies);
+    }
+
+      style['fontFamilies'] = fontFamilies;
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/text.dart]: Extra indent

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+
+  @override
+  void layout(ui.ParagraphConstraints constraints) {
+    assert(constraints.width != null);
+    // TODO(het): CanvasKit throws an exception when laid out with
+    // a font that wasn't registered.
+    try {
+      skParagraph.callMethod('layout', <double>[constraints.width]);
+    } catch (e) {
+      html.window.console.warn('CanvasKit threw an exception while laying '
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/text.dart]: This should fail harder, no?

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+
+  SkParagraphBuilder(ui.ParagraphStyle style) {
+    SkParagraphStyle skStyle = style;
+    _textDirection = skStyle._textDirection;
+    _fontFamily = skStyle._fontFamily;
+    _paragraphBuilder =
+        canvasKit['ParagraphBuilder'].callMethod('Make', <js.JsObject>[
+      skStyle.skParagraphStyle,
+      skiaFontCollection.skFontMgr,
+    ]);
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/compositor/text.dart]: nit: I think a trailing comma will allow dartfmt do a better job here.

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 4]
```diff
@@ -0,0 +1,488 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+part of engine;
+
+/// CanvasKit does not yet allow us to specify family names when registering
+/// fonts. CanvasKit reads the font name from the font's bytes. So, we map
+/// some common family names to how they are registered in the Gallery app.
+const Map<String, String> _fontFamilyOverrides = <String, String>{
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: Done.

File: lib/web_ui/lib/src/engine/compositor/text.dart [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        style['slant'] = canvasKit['FontSlant']['Italic'];
+        break;
+    }
+  }
+}
+
+class SkParagraph implements ui.Paragraph {
+  SkParagraph(this.skParagraph, this._textDirection, this._fontFamily);
+
+  final js.JsObject skParagraph;
```
[the remark this answers is not in the document]
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: Developers can hold onto it and keep using it with `Canvas.drawParagraph` if they wish. We can, however, clean up `ParagraphBuilder` after `build` is called.

File: lib/web_ui/lib/src/engine/compositor/text.dart (hunk 2 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: Changed to rethrow

File: lib/web_ui/lib/src/engine/compositor/text.dart (hunk 3 above)
Comment (hterkelsen) [reply lib/web_ui/lib/src/engine/compositor/text.dart]: Done.

**Pull request merged**